File: src/leetcode/LRUCache.java
package leetcode;

import java.util.HashMap;

public class LRUCache {
    HashMap<Integer, CacheNode> cache;
    int capacity;
    CacheNode head, last;
     
    public LRUCache(int capacity) {
        cache = new HashMap<Integer, CacheNode>();
        this.capacity = capacity;
     }
    
    public int get(int key) {
        if(cache.containsKey(key)){
            CacheNode node = cache.get(key);
            int value = node.val;
            if(node!=last){
                remove(node);
                attach(new CacheNode(key, value));
            }
            return value;
        }else{
            return -1;
        }
    }
    
    public void set(int key, int value) {
        if(cache.containsKey(key)){
            get(key);
            last.val = value;
         }else{
            CacheNode node = new CacheNode(key, value);
            attach(node);
        }
        
        if(cache.size() > capacity){
            remove(head);
        }
    }
    
    
    void remove(CacheNode node){
         if(node == head){
             cache.remove(head.key);
             head = head.next;
             return;
         }
         if(node.next != null){
             cache.remove(node.key);
             CacheNode next = node.next;
             node.key = next.key;
             node.val = next.val;
             node.next = next.next;
             if(next == last){
                 last = node;
             }
             cache.put(node.key, node);
         }             
    }
    
    void attach(CacheNode node){
        if(head == null){
            head = node;
            last = node;
        }else{
            last.next = node;
            last = last.next;
        }
        cache.put(node.key, node);
    }
    
 }

class CacheNode{
    int key;
    int val;
    CacheNode next;
    public CacheNode(int key, int val){
        this.key = key;
        this.val = val;
        next = null;
    }
}

File: python/leetcode/LRUCache.py
class LRUCache:
     
    
    # @param capacity, an integer
    def __init__(self, capacity):
        self.capacity = capacity
        self.head = None
        self.last = None
        self.cache = {}
        
    # @return an integer
    def get(self, key):
        if key in self.cache:
            node = self.cache[key]
            if node is not self.last:
                self.detach(node)
                self.attach(node)            
            return node.value
        else:
            return -1
            
    def detach(self, node):
        del self.cache[node.key]
        if node == self.head:
            self.head = node.next
            if node.next != None:
                node.next.prev = None
        else:
            prev = node.prev
            next = node.next
            prev.next = next
            next.prev = prev
            
    def attach(self,node):
        self.cache[node.key] = node
        if self.head is None:
            self.head = node
            self.last = node
        else:
            self.last.next = node
            node.prev = self.last
            self.last = self.last.next

        
        
    # @param key, an integer
    # @param value, an integer
    # @return nothing
    def set(self, key, value):
        if key in self.cache:
            self.get(key)
            self.last.value = value
        else:
            node = CacheEntity(key, value)
            self.attach(node)
        if len(self.cache) > self.capacity:
            if self.head is not None:
                self.detach(self.head)
    

class CacheEntity:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None
        self.prev = None